src/fsharp/PrettyNaming.fs

```diff
@@ -177,59 +269,89 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
-          || cat = UnicodeCategory.ConnectorPunctuation // Connectors
-          || cat = UnicodeCategory.NonSpacingMark // Combiners
-          || cat = UnicodeCategory.SpacingCombiningMark
-          || c = '\'' // Tick
-        )
-
-    /// Is this character a part of a long identifier 
-    let IsLongIdentifierPartCharacter c = 
-        (IsIdentifierPartCharacter c) || (c = '.')
-
-    let IsValidPrefixOperatorUse s = 
+        if c = '\'' then true   // Tick
+        else
+            match System.Char.GetUnicodeCategory c with
+            // Letters
+            | UnicodeCategory.UppercaseLetter
+            | UnicodeCategory.LowercaseLetter
+            | UnicodeCategory.TitlecaseLetter
+            | UnicodeCategory.ModifierLetter
+            | UnicodeCategory.OtherLetter
+            | UnicodeCategory.LetterNumber
+            // Numbers
+            | UnicodeCategory.DecimalDigitNumber
+            // Connectors
+            | UnicodeCategory.ConnectorPunctuation
+            // Combiners
+            | UnicodeCategory.NonSpacingMark
+            | UnicodeCategory.SpacingCombiningMark -> true
+            | _ -> false
+
+    /// Is this character a part of a long identifier?
+    let IsLongIdentifierPartCharacter c =
+        c = '.'
+        || IsIdentifierPartCharacter c
+
+    let IsValidPrefixOperatorUse s =
+        if System.String.IsNullOrEmpty s then false else
         match s with 
         | "?+" | "?-" | "+" | "-" | "+." | "-." | "%" | "%%" | "&" | "&&" -> true
-        | _ -> s.[0] = '!' || (s.[0] = '~' && String.forall (fun c -> c = s.[0]) s)
+        | _ ->
+            s.[0] = '!'
+            || (String.forall (fun c -> c = '~') s)
     
     let IsValidPrefixOperatorDefinitionName s = 
+        if System.String.IsNullOrEmpty s then false else
         match s with 
         | "~?+" | "~?-" | "~+" | "~-" | "~+." | "~-." | "~%" | "~%%" | "~&" | "~&&" -> true
-        | _ -> (s.[0] = '!' && s <> "!=") || (s.[0] = '~' && String.forall (fun c -> c = s.[0]) s)
+        | _ ->
+            (s.[0] = '!' && s <> "!=")
+            || (String.forall (fun c -> c = '~') s)
         
-    let IsPrefixOperator s = 
+    let IsPrefixOperator s =
+        if System.String.IsNullOrEmpty s then false else
         let s = DecompileOpName s
         match s with 
         | "~?+" | "~?-" | "~+" | "~-" | "~+." | "~-." | "~%" | "~%%" | "~&" | "~&&" -> true
-        | _ -> (s.[0] = '!' && s <> "!=")  || (s.[0] = '~' && String.forall (fun c -> c = s.[0]) s)
+        | _ ->
+            (s.[0] = '!' && s <> "!=")
+            || (String.forall (fun c -> c = '~') s)
 
     let IsTernaryOperator s = 
-        DecompileOpName s = "?<-"
-
-    let IsInfixOperator s (* where s is assumed to be a compiled name *) =
+        DecompileOpName s = qmarkSet
+
+    let IsInfixOperator =
+        /// EQUALS, INFIX_COMPARE_OP, LESS, GREATER
+        let relational = [| "=";"!=";"<";">";"$"|]
```

Comment:
The compiler generally doesn't use the ``let f = let ... in fun x -> ...`` style.  There's no real reason why not except we generally hoist the state and/or amortized values to private module bindings.  I think it's best to hoist them explicitly here too even if they are not mutable